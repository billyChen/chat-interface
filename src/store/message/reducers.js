import { createSlice } from '@reduxjs/toolkit'

const messagesSlice = createSlice({
    name: 'messages',
    initialState: [],
    reducers: {
        addMessage: {
            reducer(state, action) {
                const { message, isPublic } = action.payload

                state.push({
                    message,
                    isPublic,
                    date: Date.now()
                })
            },
            prepare(params) {
                const { message, isPublic } = params
                return { payload: { message, isPublic } }
            }
        },
        updateMessage: {
            reducer(state, action) {

            },
            prepare(params) {
                return { payload: {} }
            }
        },
        deleteMessage: {
            reducer(state, action) {

            },
            prepare(params) {
                return { payload: {} }
            }
        }
    }
})

export const { addMessage } = messagesSlice.actions

export default messagesSlice.reducer