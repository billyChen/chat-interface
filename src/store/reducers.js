import { combineReducers } from "redux";
import MessageReducer from './message/reducers'



export default combineReducers(
    {
        messages: MessageReducer
    }
)   