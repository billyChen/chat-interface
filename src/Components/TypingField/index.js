import React, { useState } from 'react'
import { useDispatch } from 'react-redux'
import clsx from 'clsx'

// Material-ui
import { makeStyles } from '@material-ui/core/styles'
import Box from '@material-ui/core/Box'
import OutlinedInput from '@material-ui/core/OutlinedInput';
import InputAdornment from '@material-ui/core/InputAdornment';
import InputLabel from '@material-ui/core/InputLabel'
import FormControl from '@material-ui/core/FormControl'
import IconButton from '@material-ui/core/IconButton'
import Button from '@material-ui/core/Button'
import VisibilityIcon from '@material-ui/icons/Visibility';
import VisibilityOffIcon from '@material-ui/icons/VisibilityOff';

// Actions
import { addMessage } from '../../store/message/reducers'


const useStyles = makeStyles({
    root: {

    },
    inputText: {
        width: '80%',
        marginTop: '15px'
    },
    button: {
        height: '50%',
        marginLeft: '15px',
        fontWeight: '600'
    },
    isDisabled: {
        color: '#cad1d9',
        pointerEvents: 'none',
        cursor: 'not-allowed',
        backgroundColor: '#f4f6f7',
    }
})

const TypingField = () => {

    const classes = useStyles()
    const dispatch = useDispatch()

    const [message, setMessage] = useState('')
    const [isPublic, setIsPublic] = useState('public')

    const handleTextChange = (e) => {
        setMessage(e.target.value)
    }

    const handleSendMessage = () => {
        dispatch(addMessage({ message, isPublic }))
        setMessage('')
    }

    const handleSetVisibility = () => {
        setIsPublic(!isPublic)
    }

    return (
        <Box display="flex" alignItems="baseline">
            <FormControl fullWidth className={clsx(classes.margin, classes.textField)} variant="outlined">
                <InputLabel htmlFor="outlined-adornment-message">Votre message</InputLabel>
                <OutlinedInput
                    data-testid="input-message"
                    id="outlined-adornment-message"
                    onChange={handleTextChange}
                    className={classes.inputText}
                    label="Search field"
                    type="search"
                    variant="outlined"
                    value={message}
                    endAdornment={
                        <InputAdornment position="end">
                            <IconButton onClick={handleSetVisibility} edge="end">
                                {isPublic ? <VisibilityIcon /> : <VisibilityOffIcon />}
                            </IconButton>
                        </InputAdornment>
                    }
                />
            </FormControl>
            <Button id="submit-button" onClick={handleSendMessage} className={clsx(classes.button, {
                [classes.isDisabled]: message.length === 0
            })} variant="contained">Envoyer</Button>
        </Box >
    )
}

export default TypingField