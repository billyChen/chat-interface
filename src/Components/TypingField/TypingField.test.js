import React from 'react'
import { Provider } from 'react-redux'
import TypingField from '.'
import configureStore from 'redux-mock-store'
import renderer from 'react-test-renderer'

// Actions
import { addMessage } from '../../store/message/reducers'
const mockStore = configureStore([]);

describe('Typing field component', () => {
    let component;
    let store;
    beforeEach(() => {
        store = mockStore({
            messages: []
        });
        store.dispatch = jest.fn();
        component = renderer.create(
            <Provider store={store}>
                <TypingField />
            </Provider>
        );
    })

    it('should render correctly', () => {
        expect(component).toMatchSnapshot()
    })

    it('should dispatch an action on click', () => {
        renderer.act(() => {
            component.root.findAllByType('button')[1].props.onClick();
        });

        expect(store.dispatch).toHaveBeenCalledTimes(1);
        expect(store.dispatch).toHaveBeenCalledWith(
            addMessage({
                isPublic: true,
                message: ''
            })
        );
    })

})