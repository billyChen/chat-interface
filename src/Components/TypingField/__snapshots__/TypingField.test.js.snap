// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typing field component should render correctly 1`] = `
<div
  className="MuiBox-root MuiBox-root-5"
>
  <div
    className="MuiFormControl-root MuiFormControl-fullWidth"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
      data-shrink={false}
      htmlFor="outlined-adornment-message"
    >
      Votre message
    </label>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root makeStyles-inputText-2 MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
      data-testid="input-message"
      onClick={[Function]}
      variant="outlined"
    >
      <input
        aria-invalid={false}
        className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd MuiInputBase-inputTypeSearch"
        disabled={false}
        id="outlined-adornment-message"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="search"
        value=""
      />
      <div
        className="MuiInputAdornment-root MuiInputAdornment-positionEnd"
      >
        <button
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
              />
            </svg>
          </span>
        </button>
      </div>
      <fieldset
        aria-hidden={true}
        className="PrivateNotchedOutline-root-6 MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="PrivateNotchedOutline-legendLabelled-8"
        >
          <span>
            Search field
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-3 makeStyles-isDisabled-4"
    disabled={false}
    id="submit-button"
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      Envoyer
    </span>
  </button>
</div>
`;
